### ISSUE TITLE
Assembler queue access with ingame script

### ISSUE THREAD

**Issue by mexmer** [KeenSoftwareHouse/SpaceEngineers#96]
One of this ingame PB pb is missing is access to assembler queue.

It seems not hard to make current queue visible, but i'm not sure if it will be safe (well i'm pretty sure, it will not), therefore queue will need to be accessed either by functions (addtoqueue, removefrom queue), or using interface (IMyQueue), like is for IMyInvetory.

I'm more inclined to IMyQueue (or whatever we will call it) as safe access to queue

ideas welcome

**Comment by mrsib** [comment]
I was thinking it didn't need to be too complicated. Just add items using their type/subtype name/amount and a function to clear the queue:
`myAssembler.QueueItem("MyObjectBuilder_Component", "SmallTube", 100));
myAssembler.ClearQueue();`

**Comment by MMaster** [comment]
I think PB also needs access to read the queue so scripts can figure out what is being produced. So IMyQueue approach would be probably better.

**Comment by mexmer** [comment]
@MMaster that's my thought about it. I want safe exposure of queue, so we can do real scripts for watching production and queuing production.

**Comment by MMaster** [comment]
@mexmer Definitely see your point - it's really needed for production automation scripts. I am really happy that there is so great modding community around SE. Half of the things I wanted to address already have pull requests or issues opened :)

**Comment by mrsib** [comment]
@mexmer @MMaster Yes I think it would be good to have access to the queue, I have created a merge for the basic functioned I mentioned as it was very easy to do.
The problem with the Queue is that it uses MyBlueprintDefinitionBase which isn't really exposed to InGame programming. 
What I am thinking is a copy of the Queue where instead of MyBlueprintDefinitionBase, there is type and subtypename. This would mean defining a struct (exposed to InGame), and returning a list with that struct:
`GetQueue(); // returns List<AssemblerQueueItem>`
Then queue removal would be easier, just expose a function to remove at index. i.e. `RemoveQueueItem(int itemIndex);`

**Comment by mexmer** [comment]
@mrsib problem with this list is, that is unsafe. i was already doing some stuff with assembler in my local branch, and there are lot of issues involving sync and safe operation, therefore i would not be hasty with implementing proprietary functions, that might get removed soon (more likely will be obsoleted)

**Comment by mrsib** [comment]
@mexmer Ok, 
Can you help me understand why the list idea is unsafe and what sync issues it could cause.
Also, other than failure to use if checks in my previous commit, was there any issues with safety/sync?
I admit I was a bit hasty on the commit, I am pretty excited to be able to work on this project.

**Comment by mexmer** [comment]
List<AssemblerQueueItem> itself is unsafe, because it's not synced, only operations that add, remove from it do locking to prevent problems. What you should do is most likely copy queue over. I was looking at IMyInventory interface, and atm. i'm not sure, what will be best and safest approach.
Main problem is, that on working assembler queue continuously updates, it sends sync messages to ui, but it can't send sync message to pb.

atm. i'm working on powerconsumer interface, i plan to do further invenstigation in production queue later. i have it on my todo list since source came out ... just not enough time, that's why i've put issue up, so anyone else can look into it.

**Comment by mrsib** [comment]
I understand the issue. But what I am thinking is if you take a copy of the queue and act upon it during PB execution (and don't store it for later) then it will be up to date as the PB runs in the game loop and the assembler queue wont change state.
Worse case scenario, is someone saves the AssemblerQueueItem they want to remove, but when PB is next run it is not there. 
So perhaps instead of using
`RemoveQueueItem(int itemIndex);`
we could use:
`RemoveQueueItem(AssemblerQueueItem queueItem);`
Which in turn checks the item at index has same type/subtypename/amount, then removes item and returns true, otherwise returns false.

**Comment by mrsib** [comment]
@mexmer @MMaster
I carried on and committed to my local branch the full solution that was in my head.
https://github.com/mrsib/SpaceEngineers/commit/4059e16eb30b4b55ffa7954174daae3190ca707d

I believe as I am not accessing m_queue directly when modifying the queue that this should sync and be safe. I use the MyProductionBlock functions `InsertQueueItemRequest` and `RemoveQueueItemRequest`

**Comment by Drakoskai** [comment]
I took a look at your commit, basically your list is a list of references to the items in the the assembler queue.  Let's call it a "critical resource".  If multiple actors modify or read the list simultaneously, the order of those reads and writes will not be guaranteed and you will have a Data Race to see which takes effect first. The solution is to use one of the Concurrent Collections to synchronize the access so that all actors seeking to use the critical resource are put in order.  Be careful --  The concurrency that is currently in place uses algorithms with overhead so the price you pay for synchronization is latency. Also keep in mind that when you access the collection and lock it for your use, you are blocking all other threads access to it.

I hope that helps :)

**Comment by mrsib** [comment]
I understand the concurrency issues, however I feel that it is not necessary to use the extra overhead in this case.
If two actors simultaneously add item to the queue, there will be two items in the queue. I.e. PB adds 100 steel plate, user adds 100 steel plate. Now either the user can realise the mistake and remove his 100, or the PB can check there only the amount needed in the queue periodically and correct if needed. If neither then there's 200 Steel plate produced.
If two actors simultaneously remove item from the queue, i.e. user removes one through U.I and PB removes one, one will fail gracefully, the other will succeed and remove the item. The PB has a return value so can detect failure, recheck queue and retry if the item still needs removing.
Basically I don't think the list is a 'critical resource'. But I could be wrong, I haven't worked with many multi-threaded applications, so please if I am wrong, educate me :) .

**Comment by Drakoskai** [comment]
The problem in technical terms is that each thread will make a local copy of the resource and when you "write" a change, it will store it to the local copy and not to the shared resource.  In order to write it to the shared resource you have to do a "store/store" things get hairy here because we're getting into the realm of memory models and also the fact that the JIT will re-order instructions to optimize and that may optimize you into a race condition.  

http://preshing.com/20120710/memory-barriers-are-like-source-control-operations/

**Comment by mrsib** [comment]
Ok well the list is meant to be a snapshot and not to be modified. The add and remove item functions are for modifying internal queue and use internal methods that handle sync/concurrency issues. Perhaps to make this clear in the code the GetQueue function could be called GetQueueSnapshot? Or am I still missing the point?

**Comment by mrsib** [comment]
No further advice, and from testing (albiet limited), reading further into this, and looking at other in game methods such is IMyInventory, there doesn't appear to be problems with this method.
I have made a few changes, but I think its ready for a pull request:
https://github.com/KeenSoftwareHouse/SpaceEngineers/compare/master...mrsib:Sibz-Assembler-Functions
Added functions:
GetQueueItems - Returns list of items in queue with string format type/subtypename
AddQueueItem - Adds item to queue using string type/subtypename and int amount
RemoveQueueItem - Removes an item from the queue
ClearQueue - Clears all items from queue
GetQueueItemAmount - Returns total amount of item in queue by subtypename

**Comment by MMaster** [comment]
I see that GetQueueItemAmount takes only subtype name. I think there is type missing as there can be items with the same subtype name but different type. Other than that I think it looks good.

**Comment by mrsib** [comment]
https://github.com/KeenSoftwareHouse/SpaceEngineers/commit/ac0a964c4ed7870b6f2aa02ff7034b0dd9175bae

**Comment by Sibz** [comment]
Thinking long and hard about this and testing further, my solution is not up to scratch. It seems the values in the list get messed up when accessing them in game. Debugging it shows me correct values returned from GetQueueItems(). But accessing them in game causes issue. It works as far as passing an item off the list back to RemoveQueueItem(), but idx/amount are unexpectedly large numbers, and accessing the strings causes an exception.

**Comment by Sibz** [comment]
Tried and failed to implement an IMyAssemblerQueue interface.

I am thinking that it would be easier to add just functions:
```c#
/// <summary>
        /// Adds an item to the assembler queue
        /// </summary>
        /// <param name="itemType">type of item, i.e. 'MyObjectBuilder_Component'</param>
        /// <param name="subtypeName">Shortname of component, i.e. SteelPlate, Reactor, MetalGrid, etc.</param>
        /// <param name="amount">Quantity of specified item to add to queue</param>
        /// <returns>
        /// true - item added successfully
        /// false - item not added - probably because itemType or subtypeName are invalid
        /// </returns>
        bool AddQueueItem(string itemType, string subtypeName, int amount);

        /// <summary>
        /// Retrieves the number of items in the assembler queue
        /// </summary>
        /// <returns></returns>
        int GetQueueCount();

        /// <summary>
        /// Retrieves an item at index
        /// </summary>
        /// <param name="idx"></param>
        /// <returns>String with ItemId and SubtypeName<br />
        /// i.e. "MyObjectBuilder_Component\SteelPlate"
        /// </returns>
        string GetQueueItemAt(int idx);

        /// <summary>
        /// Removes all matching queue items
        /// </summary>
        /// <param name="itemType"></param>
        /// <param name="subtypeName"></param>
        /// <returns></returns>
        bool RemoveQueueItem(string itemType, string subtypeName);

        /// <summary>
        /// Removes a queue item at specific index
        /// </summary>
        /// <param name="idx"></param>
        /// <returns></returns>
        bool RemoveQueueItemAt(int idx);

        /// <summary>
        /// Removes all items in the assembler queue
        /// </summary>
        void ClearQueue();
        
        /// <summary>
        /// Returns the total amount of an item in the queue
        /// </summary>
        /// <param name="itemType">the item type of item, i.e. MyObjectBuilder_Component</param>
        /// <param name="subtypeName">The subtype name of item to count, i.e. SteelPlate</param>
        /// <returns>int - amount queued in total</returns>
        int GetQueueItemAmount(string itemType, string subtypeName);

        /// <summary>
        /// Retrieves count of items at an index
        /// </summary>
        /// <param name="idx"></param>
        /// <returns></returns>
        int GetQueueItemAmountAt(int idx);
`
